desktop/cmp/treemap/TreeMap.js

```diff
@@ -0,0 +1,317 @@
+                }
+            },
+            series: [{
+                data,
+                type: 'treemap',
+                animation: false,
+                layoutAlgorithm: algorithm,
+                borderWidth: 0,
+                turboThreshold: maxNodes,
+                dataLabels: {
+                    enabled: true,
+                    allowOverlap: false,
+                    align: 'left',
+                    verticalAlign: 'top',
+                    style: {textOutline: 'none', visibility: 'hidden'}
+                },
+                events: {click: this.onClick}
+            }]
+        };
+    }
+
+    //----------------------
+    // Click handling
+    //----------------------
+    onClick = (e) => {
+        this._clickCount++;
+        this._debouncedClickHandler(e.point.record, e);
+        if (this._clickCount >= 2) this._debouncedClickHandler.flush();
+    };
+
+    clickHandler(record, e) {
+        try {
+            const {onClick, onDoubleClick} = this.model;
+            if (onClick && this._clickCount === 1) {
+                onClick(record, e);
+            } else if (onDoubleClick) {
+                onDoubleClick(record, e);
+            }
+        } finally {
+            this._clickCount = 0;
+        }
+    }
+
+    //----------------------
+    // Selection handling
+    //----------------------
+    syncSelection() {
+        if (!this._chart) return;
+
+        const {selectedIds, maxDepth, gridModel, store} = this.model;
+
+        // Fallback to parent node if selection exceeds max depth
+        let toSelect;
+        if (maxDepth && gridModel && gridModel.treeMode) {
+            toSelect = new Set(selectedIds.map(id => {
+                const record = store.getById(id);
+                return record ? record.xhTreePath.slice(0, maxDepth).pop() : null;
+            }));
+        } else {
+            toSelect = new Set(selectedIds);
+        }
+
+        // Update selection in chart
+        this._chart.series[0].data.forEach(node => {
+            node.select(toSelect.has(node.id), true);
+        });
+    }
+
+    //----------------------
+    // Labels
+    //----------------------
+    async updateLabelVisibilityAsync() {
+        if (!this._chart) return;
+
+        // Show / hide labels by comparing label size to node size
+        await forEachAsync(this._chart.series[0].data, node => {
+            if (node.dataLabel && node.graphic) {
+                const buffer = 10,
+                    tooSmallWidth = (node.dataLabel.width + buffer) > node.graphic.element.width.baseVal.value,
+                    tooSmallHeight = (node.dataLabel.height + buffer) > node.graphic.element.height.baseVal.value,
+                    style = tooSmallWidth || tooSmallHeight ? {visibility: 'hidden'} : {visibility: 'visible'};
+
+                node.update({dataLabels: {style}}, false, false);
+            }
+        });
+
+        if (this._chart) this._chart.redraw();
```

Comment:
Here's another place where I think we could be a bit more defensive to avoid extra calls to `redraw()`. 

Suggest we check to see if we actually need to make these updates - looks like the node's current visibility can be accessed via `node.dataLabel.styles.visibility`.  If we need to change, we can call `node.update()` and mark a flag set outside the loop to indicate that the loop actually did something. We can then check that before calling `redraw()`.

With a bit of logging and profiling it appears as though this will save us real-world work for some updates.